tools/pre_ci.py

```
@@ -395,49 +401,92 @@ def execute(self) -> None:
 
             # 4. Unit Tests (Run sequentially to see clear progress)
-            # Pytest executes in the remote test-matrix job. Prevent redundancy in CI.
             if not self.is_ci:
-                self.run_command(
-                    ["uv", "run", "--no-sync", "pytest", "-v"],
-                    "Unit Tests & Coverage Enforcement",
+                print("\n>>> [Step: Unit Tests & Coverage Enforcement]", flush=True)
+                # Stream output directly to console for real-time visibility.
+                # Coverage failure is detected via pytest's own exit code when
+                # --cov-fail-under is passed correctly in pyproject.toml.
+                res = subprocess.run(
```

Ruff (0.15.12)

[error] `subprocess` call: check for execution of untrusted input

(S603)